runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1174,23 +1029,400 @@ static iree_status_t iree_hal_hip_device_queue_write(
+
+  iree_hal_command_buffer_t* command_buffer = data->command_buffer;
+  if (iree_hal_hip_multi_queue_command_buffer_isa(command_buffer)) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_hip_multi_queue_command_buffer_get(
+                command_buffer, data->queue_affinity, &command_buffer));
+  }
+
+  if (iree_hal_hip_stream_command_buffer_isa(command_buffer)) {
+    status = iree_status_join(
+        status,
+        iree_hal_stream_tracing_context_collect_list(
+            // Get the tracing context from the device/stream/queue affinity.
+            device->topology.devices[idx].tracing_context,
+            // Get the tracing event list from the command buffer.
+            iree_hal_hip_stream_command_buffer_tracing_events(command_buffer)
+                .head));
+  } else if (iree_hal_hip_graph_command_buffer_isa(command_buffer)) {
+    status = iree_status_join(
+        status,
+        iree_hal_stream_tracing_context_collect_list(
+            // Get the tracing context from the device/stream/queue affinity.
+            device->topology.devices[idx].tracing_context,
+            // Get the tracing event list from the command buffer.
+            iree_hal_hip_graph_command_buffer_tracing_events(command_buffer)
+                .head));
+  }
+
+  // 2) Free the event we specifically created here.
+  iree_hal_hip_event_release(event);
+
+  // 3) Notify all of the signal semaphores that they have been incremented.
+  for (iree_host_size_t i = 0; i < data->signal_semaphore_list.count; ++i) {
+    uint64_t unused_return_value = 0;
+    // We use query to force the semaphore to update.
+    iree_status_ignore(iree_hal_semaphore_query(
+        data->signal_semaphore_list.semaphores[i], &unused_return_value));
+  }
+
+  // 4) Free the iree_hal_hip_device_semaphore_submit_callback_data_t and
+  // the resource set attached.
+  iree_hal_resource_set_free(data->resource_set);
+  iree_slim_mutex_deinitialize(&data->status_mutex);
+  iree_allocator_free(device->host_allocator, data);
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static iree_status_t iree_hal_hip_device_execute_now(
+    iree_hal_hip_device_semaphore_submit_callback_data_t* data) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_ASSERT_EQ(iree_math_count_ones_u64(data->queue_affinity), 1,
+                 "Cannot execute a command buffer on more than one queue");
+  iree_hal_hip_device_t* device = data->device;
+  iree_status_t status = iree_ok_status();
+
+  // If we had a semaphore failure, then we should propagate it
+  // but not run anything
+  if (!iree_status_is_ok(data->status)) {
+    for (iree_host_size_t i = 0; i < data->signal_semaphore_list.count; ++i) {
+      iree_hal_semaphore_fail(data->signal_semaphore_list.semaphores[i],
+                              iree_status_clone(data->status));
+    }
+    return data->status;
+  }
+
+  uint32_t idx = iree_math_count_trailing_zeros_u64(data->queue_affinity);
```

Comment:
int (consistency is key, the type is int but you've used uint32_t and uint64_t - this is confusing to a reader as it implies that there's a specific reason you used a different type)